### PR TITLE
Implement server stop command for background instances

### LINKED ISSUE #454: Stop server
Add command line support to stop CodeChecker server if it was started in the background.
A user should be able to list the servers started by him.

### PR DESCRIPTION
**Pull request by whisperity**
* Servers now register and deregister themselves for the running user's
   instance list (stored in: `/tmp/.codechecker_username.instances.json`)

 * New commands under `CodeChecker server`:
   * `-l`/`--list` lists the running instances for the user
   * `-s`/`--stop` stops the server for the given `--workspace` and
     `--view-port` combination
   * `--stop-all` stops every running CodeChecker instance for the user

 * Servers stopped by classical SIGTERM unregister themselves, but the
   instance list is always double-checked to ensure that only the actual
   running processes are shown.

This feature introduces _psutil_ as a new Python module dependency.
This closes #454.

### CODE REVIEW

File: codechecker_lib/arg_handler.py [hunk 1]
```diff
@@ -101,6 +103,40 @@ def handle_server(args):
 
     workspace = args.workspace
 
+    if (args.list or args.stop or args.stop_all) and \
+            not (args.list ^ args.stop ^ args.stop_all):
+        print("CodeChecker server: error: argument -l/--list and -s/--stop "
```
Comment (Xazax-hun) [inline codechecker_lib/arg_handler.py]: I know this is not consistent at the moment, but we might want to avoid print and use LOG.error here.

File: codechecker_lib/instance_manager.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+def __getInstanceDescriptorPath():
+    return os.path.join(tempfile.gettempdir(), ".codechecker_" +
+                        getpass.getuser() + ".instances.json")
+
+
+def __makeInstanceDescriptorFile():
+    if not os.path.exists(__getInstanceDescriptorPath()):
+        with open(__getInstanceDescriptorPath(), 'w') as f:
+            json.dump([], f)
+        os.chmod(__getInstanceDescriptorPath(), stat.S_IRUSR | stat.S_IWUSR)
```
Comment (Xazax-hun) [inline codechecker_lib/instance_manager.py]: Maybe the code is easier to understand if you only call the `getInstanceDescriptorPath` once.

File: codechecker_lib/util.py [hunk 3]
```diff
@@ -188,9 +190,56 @@ def call_command(command, env=None):
...
+    proc.terminate()
+
+    # If children processes don't stop gracefully in time,
+    # slaughter them by force.
+    __, still_alive = psutil.wait_procs(children, timeout=5)
+    for p in still_alive:
+        p.kill()
+
+
+# -------------------------------------------------------------------------
```
Comment (Xazax-hun) [inline codechecker_lib/util.py]: I am in favour of removing these lines. Just a suggestion though.

File: codechecker_lib/util.py [hunk 4]
```diff
@@ -188,9 +190,56 @@ def call_command(command, env=None):
...
     Default workspace in the users home directory.
     """
     workspace = os.path.join(os.path.expanduser("~"), '.codechecker')
     return workspace
+
+
+# -------------------------------------------------------------------------
+def print_table(lines, separate_head=True):
+    """Prints a formatted table given a 2 dimensional array."""
+    # Count the column width.
```
Comment (Xazax-hun) [inline codechecker_lib/util.py]: Empty line between the commend and the corresponding code.

File: codechecker_lib/arg_handler.py (hunk 1 above)
Comment (whisperity) [inline codechecker_lib/arg_handler.py]: I tried to be consistent with the output of `argparse`.

Comment (Xazax-hun) [inline codechecker_lib/arg_handler.py]: Oh, I see. Its ok then.

**Comment by gyorb** [comment]
Can we store the name of the hosts? It is possible that the file system is mounted to multiple hosts and it might be useful at least to see all of the CodeCheckers server started by the user, even if stopping instances on other servers is not possible.

**Comment by gyorb** [comment]
This feature should be merged after the 5.7.1 bugfix release is out.

File: codechecker/CodeChecker.py [hunk 5]
```diff
@@ -53,7 +53,7 @@ def __call__(self, parser, namespace, value, option_string=None):
         namespace.ordered_checkers = ordered_checkers
 
 
-# ------------------------------------------------------------------------------
+# -----------------------------------------------------------------------------
```
Comment (Xazax-hun) [inline codechecker/CodeChecker.py]: Maybe we could remove these as well if we are already touching these.

File: codechecker_lib/arg_handler.py [hunk 6]
```diff
@@ -167,7 +221,7 @@ def handle_server(args):
                                              db_connection_string,
                                              suppress_handler,
                                              args.not_host_only,
-                                             context.db_version_info)
+                                             context)
```
Comment (gyorb) [inline codechecker_lib/arg_handler.py]: Do we really need the whole context object? I try to minimize the usage of it. It is a global object with a lots of information. If not really required I would suggest not to pass it as an argument.

File: tests/functional/package_test/__init__.py [hunk 7]
```diff
@@ -37,6 +37,8 @@
 # Stopping event for CodeChecker server.
 __STOP_SERVER = multiprocessing.Event()
 
+__shared = {}
```
Comment (gyorb) [inline tests/functional/package_test/__init__.py]: The test infrastructure is completely rewritten, with different test layout/setup and configuration. For easier merge maybe we should split the testing part into a separate pull request.

File: viewer_server/client_db_access_server.py [hunk 8]
```diff
@@ -312,7 +319,7 @@ def process_request(self, request, client_address):
 
 
 def start_server(package_data, port, db_conn_string, suppress_handler,
-                 not_host_only, db_version_info):
+                 not_host_only, context):
```
Comment (gyorb) [inline viewer_server/client_db_access_server.py]: Do we use the context here only to get the `db_version_info` ?

Comment (whisperity) [inline viewer_server/client_db_access_server.py]: No, the context carries the `codechecker_workspace` information into the server.

**Comment by gyorb** [comment]
Test should be added #587
